rfcs/0008-over-window.md

```diff
@@ -0,0 +1,126 @@
+
+[window_function_wiki]: https://en.wikipedia.org/wiki/Window_function_(SQL)#:~:text=In%20SQL%2C%20a%20window%20function,single%20value%20for%20multiple%20rows.
+
+## Definitions
+
+Window function call:
+
+```plain
+function_name ([expression [, expression ... ]]) [ FILTER ( WHERE filter_clause ) ] OVER window_name
+function_name ([expression [, expression ... ]]) [ FILTER ( WHERE filter_clause ) ] OVER ( window_definition )
+function_name ( * ) [ FILTER ( WHERE filter_clause ) ] OVER window_name
+function_name ( * ) [ FILTER ( WHERE filter_clause ) ] OVER ( window_definition )
+```
+
+Window definition:
+
+```plain
+[ existing_window_name ]
+[ PARTITION BY expression [, ...] ]
+[ ORDER BY expression [ ASC | DESC | USING operator ] [ NULLS { FIRST | LAST } ] [, ...] ]
+[ frame_clause ]
+```
+
+frame_cause:
+
+```plain
+ RANGE | ROWS | GROUPS } frame_start [ frame_exclusion ]
+{ RANGE | ROWS | GROUPS } BETWEEN frame_start AND frame_end [ frame_exclusion ]
+```
+
+We can simplify that in our first version:
+
+```plain
+[ existing_window_name ]
+[ PARTITION BY expression ]
+[ ORDER BY column_ref ]
+[ frame_clause ]
+```
+
+```plain
+ROWS BETWEEN frame_start AND frame_end
+```
+
+## Motivation
+
+The feature is a part of the SQL standard and is widely supported by different SQL databases and streaming systems.
+
+The feature is also helpful in fraud detection.
+
+## Design
+
+### Primary key derivation
+
+The output rows count of OverWindow is exact as the input rows, as is the primary key.
+
+### Distribution key
+
+In most window definitions, one partition key is specified, and the data should be exchanged by the partition key.
+
+### BatchOverWindow
+
+The implementation is trivial, the only thing we need to concern is that the batch aggregator also needs to support `retract`, which is much more similar to the stream aggregator.
+
+### StreamOverWindow
+
+The design highly depends on [the WatermarkFilter and StreamSort operator](https://github.com/risingwavelabs/rfcs/pull/1).
+
+If the input stream is ordered, then the implementation is trivial.
+
+There are some limitations in stream aggregation:
+
+* If the sort key is specified, then a watermark must exist on the column.
+* If no sort key is specified, then the result is undefined but may be helpful; it's roughly the same as the proctime in Flink.
+* `FOLLOWING` is **banned; the max value of `frame_start` and `frame_end` is `CURRENT ROW`.
+* Window functions that peek the future rows are **banned**, e.g., `lag`.
+
+### Multiple windows
+
+In the SQL standard, different windows are allowed in the queries, although there are few cases to use that.
+
+We can support them in the frontend easily.
+
+Assume we have two windows:
+
+```plain
+WINDOW w1 AS (PARTITION BY p1 ORDER BY s1)
+WINDOW w2 AS (PARTITION BY p2 ORDER BY s3)
+```
+
+We can plan the query as two branches with a join to combine them:
```

Comment:
This sounds not that trivial under the current planner & fragmenter. 🤣

Reply:
May off-topic: based on a similar idea, we can also enable two-phase optimization as much as possible if we can split agg calls and join them back. 🤩

Reply:
Yes, we need DAG planner for that.